pkg/trait/toleration.go

```diff
@@ -0,0 +1,121 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trait
+
+import (
+	"fmt"
+	"strconv"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
+	"github.com/apache/camel-k/pkg/util"
+)
+
+// This trait will enable a Toleration.
+// Tolerations are applied to pods, and allow (but do not require) the pods to schedule onto nodes with matching taints.
+// See https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/ for more details.
+//
+// It's disabled by default.
+//
+// +camel-k:trait=toleration
+type tolerationTrait struct {
```

Comment:
Would there be a way to specify multiple tolerations?

Reply:
Yeah, I tried to implement that as multiple tolerations but the problem I see is that a toleration have more than 1 attribute, so, something like this
```
kamel run ... -t toleration.key=key1 -t toleration.operator=Exists -t toleration.effect=NoExecute -t toleration.key=key2 -t toleration.operator=Equal -t toleration.value=val2 -t toleration.effect=NoExecute 
```
May mix the values and it does not result user friendly imo. Maybe a regex to selfcontain a toleration could be better, like:
```
kamel run ... -t toleration.expr=key1:Exists:NoExecute -toleration.expr=key2:Equal:val2:NoExecute
```
Wdyt?

Reply:
or even better
```
kamel run ... -t toleration.expr=key1:NoExecute -toleration.expr=key2=val2:NoExecute
```
assuming that if the `=` operator exists, then it's an `Equal`, otherwise an `Exists` operator

Reply:
I agree that, currently, mapping from CLI options to trait configuration is too limited to support structuring multi-valued options, and the best approach is to take the serialized form.

Anyway, I find it also more concise, and it's what the `kubectl taint` command takes as input: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/.

So I'd be inclined toward `-t toleration.expr=key1:Exists:NoExecute`, and maybe changing `expr` by `taint`.